src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs

```diff
@@ -163,12 +177,111 @@ protected override void ExecuteCore()
+
+            IEnumerable<RuntimePackAssetInfo> runtimePackAssets =
+                RuntimePackAssets.Select(item => RuntimePackAssetInfo.FromItem(item));
+
+
+            ProjectContext projectContext = lockFile.CreateProjectContext(
+                NuGetUtils.ParseFrameworkName(TargetFramework),
+                RuntimeIdentifier,
+                PlatformLibraryName,
+                RuntimeFrameworks,
+                IsSelfContained);
+
+            var builder = new DependencyContextBuilder2(mainProject, projectContext, IncludeRuntimeFileVersions);
+
+            builder = builder
+                .WithMainProjectInDepsFile(IncludeMainProject)
+                .WithReferenceAssemblies(referenceAssemblyInfos)
+                .WithDirectReferences(directReferences)
+                .WithDependencyReferences(dependencyReferences)
+                .WithReferenceProjectInfos(referenceProjects)
+                .WithExcludeFromPublishAssets(excludeFromPublishAssets)
+                .WithRuntimePackAssets(runtimePackAssets)
+                .WithCompilationOptions(compilationOptions)
+                .WithReferenceAssembliesPath(FrameworkReferenceResolver.GetDefaultReferenceAssembliesPath())
+                .WithPackagesThatWereFiltered(GetFilteredPackages());
+
+            if (CompileReferences.Length > 0)
+            {
+                builder = builder.WithCompileReferences(ReferenceInfo.CreateReferenceInfos(CompileReferences));
+            }
+
+            var resolvedNuGetFiles = ResolvedNuGetFiles.Select(f => new ResolvedFile(f, false))
+                                .Concat(ResolvedRuntimeTargetsFiles.Select(f => new ResolvedFile(f, true)));
+            builder = builder.WithResolvedNuGetFiles(resolvedNuGetFiles);
+
+            DependencyContext dependencyContext = builder.Build();
+
+            var writer = new DependencyContextWriter();
+            using (var fileStream = File.Create(depsFilePath))
             {
                 writer.Write(dependencyContext, fileStream);
             }
-            _filesWritten.Add(new TaskItem(DepsFilePath));
+            _filesWritten.Add(new TaskItem(depsFilePath));
+        }
 
+        protected override void ExecuteCore()
+        {
+            if (DepsFileGenerationMode.Equals("old", StringComparison.InvariantCultureIgnoreCase))
+            {
+                WriteDepsFileOld(DepsFilePath);
+            }
+            else if (DepsFileGenerationMode.Equals("new", StringComparison.InvariantCultureIgnoreCase))
+            {
+                WriteDepsFileNew(DepsFilePath);
+            }
+            else
+            {
+                var newDepsFilePath = Path.ChangeExtension(DepsFilePath, ".new.json");
+
+                WriteDepsFileOld(DepsFilePath);
+
+                WriteDepsFileNew(newDepsFilePath);
+
+                var oldJson = JObject.Parse(File.ReadAllText(DepsFilePath));
+                var newJson = JObject.Parse(File.ReadAllText(newDepsFilePath));
```

Comment:
This is quite expensive. deps files can be big. We're reading into Strings unnecessarily. I know this isn't an inner loop task, but that's a lot of waste.